boot/freeldr/freeldr/arch/drivers/hwide.c

```diff
@@ -2,940 +2,1252 @@
-    TRACE("SK 0x%x, ASC 0x%x, ASCQ 0x%x\n",
-          SenseData.SenseKey,
-          SenseData.AdditionalSenseCode,
-          SenseData.AdditionalSenseCodeQualifier);
+    if (SenseData.SenseKey != SCSI_SENSE_NOT_READY)
+        return TRUE;
+
+    if (SenseData.AdditionalSenseCode == SCSI_ADSENSE_NO_MEDIA_IN_DEVICE)
+        return FALSE;
 
-    if (SenseData.SenseKey == SCSI_SENSE_NOT_READY)
+    switch (SenseData.AdditionalSenseCodeQualifier)
     {
-        if (SenseData.AdditionalSenseCode == SCSI_ADSENSE_LUN_NOT_READY)
-        {
-            switch (SenseData.AdditionalSenseCodeQualifier)
-            {
-                case SCSI_SENSEQ_BECOMING_READY:
-                    /* Wait until the CD is spun up */
-                    StallExecutionProcessor(4e6);
-                    return FALSE;
+        case SCSI_SENSEQ_BECOMING_READY:
+            /* Wait until the CD is spun up */
+            StallExecutionProcessor(4e6);
+            return FALSE;
 
-                case SCSI_SENSEQ_INIT_COMMAND_REQUIRED:
-                    /* The drive needs to be spun up, send the SCSI READ TOC command */
-                    RtlZeroMemory(&AtapiPacket, sizeof(AtapiPacket));
-                    AtapiPacket[0] = SCSIOP_READ_TOC;
-                    AtapiPacket[7] = (MAXIMUM_CDROM_SIZE << 8) & 0xFF;
-                    AtapiPacket[8] = MAXIMUM_CDROM_SIZE & 0xFF;
-                    AtapiPacket[9] = READ_TOC_FORMAT_SESSION << 6;
-                    Success = AtaSendAtapiPacket(DeviceUnit->Channel,
-                                                 AtapiPacket,
-                                                 ATAPI_PACKET_SIZE(DeviceUnit->IdentifyData),
-                                                 MAXIMUM_CDROM_SIZE);
-                    if (!Success)
-                    {
-                        AtapiPrintSenseData(DeviceUnit);
-                        return FALSE;
-                    }
+        case SCSI_SENSEQ_INIT_COMMAND_REQUIRED:
+            /* The drive needs to be spun up */
+            AtapAtapiReadToc(DeviceUnit);
+            return FALSE;
 
-                    AtaReadBuffer(DeviceUnit->Channel, &DummyData, MAXIMUM_CDROM_SIZE);
-                    /* fall through */
+        default:
+            return FALSE;
+    }
 
-                default:
-                    DeviceUnit->Flags &= ~ATA_DEVICE_NOT_READY;
-                    return FALSE;
+    return TRUE;
+}
 
-            }
-        }
-        else if (SenseData.AdditionalSenseCode == SCSI_ADSENSE_NO_MEDIA_IN_DEVICE)
-        {
-            DeviceUnit->Flags |= ATA_DEVICE_NO_MEDIA;
-            return FALSE;
-        }
+static
+BOOLEAN
+AtapAtapiInitDevice(
+    _In_ PHW_DEVICE_UNIT DeviceUnit)
+{
+    PIDENTIFY_PACKET_DATA IdentifyPacketData = &DeviceUnit->IdentifyPacketData;
+    ULONG i;
+
+    DeviceUnit->CdbSize = AtaDevCdbSizeInWords(IdentifyPacketData);
+
+    /* Clear the ATAPI 'Bus reset' indication */
+    for (i = 0; i < 5; ++i)
```

Comment:
It has to be done only 5 times? Not less and not more?

Reply:
Added a comment.